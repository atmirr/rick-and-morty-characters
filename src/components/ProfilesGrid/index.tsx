import React from 'react';
import { isEmpty } from 'lodash';
import styled from 'styled-components';
import ProfilePlaceholders from './components/ProfilePlaceholders';
import ProfileItem from './components/ProfileItem';

const Container = styled.div`
  max-width: 1400px;
  margin: 0 auto;
  padding-bottom: 40px;
`;

const Grid = styled.div`
  padding: 30px 25px;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
`;

const ErrorWrapper = styled.div`
  text-align: center;
`;

const Error = styled.span`
  font-size: 16px;
  color: #c53c3c;
`;

type PropTypes = {
  profiles: {
    id: number;
    name: string;
    species: string;
    status: string;
    gender: string;
    image: string;
    episode: string[];
    location: {
      name: string;
      url: string;
    };
  }[];
  error: string | null;
  isLoading: boolean;
};

function ProfilesGrid({
  profiles,
  error,
  isLoading,
}: PropTypes): React.ReactElement {
  const errorText = `There is something wrong! Error: ${error}`;
  return (
    <Container data-testid="profiles-grid">
      <Grid>
        {!isEmpty(profiles) &&
          profiles.map(
            ({
              id,
              name,
              species,
              status,
              gender,
              image,
              episode,
              location,
            }) => (
              <ProfileItem
                name={name}
                species={species}
                status={status}
                gender={gender}
                image={image}
                locationUrl={location?.url}
                episodeUrls={episode}
                key={id}
              />
            ),
          )}
        {isLoading && <ProfilePlaceholders count={20} />}
      </Grid>
      {error && (
        <ErrorWrapper>
          <Error>{errorText}</Error>
        </ErrorWrapper>
      )}
    </Container>
  );
}

export default ProfilesGrid;
