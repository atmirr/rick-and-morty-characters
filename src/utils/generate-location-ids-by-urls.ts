export const generateLocationIdsByUrls = (urls: string[]): string[] =>
  urls.reduce((result: string[], url: string) => {
    const splittedUrl = url.split('/');
    const id = splittedUrl[splittedUrl.length - 1];
    return id ? [...result, id] : [...result];
  }, []);
