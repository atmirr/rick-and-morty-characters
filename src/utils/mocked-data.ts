export const characters = [
  {
    id: 1,
    name: 'Rick Sanchez',
    species: 'Human',
    gender: 'Male',
    created: '2017-11-04T18:48:46.250Z',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
    ],
    image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    origin: {
      name: 'Earth (C-137)',
      url: 'https://rickandmortyapi.com/api/location/1',
    },
    status: 'Alive',
    type: '',
    url: 'https://rickandmortyapi.com/api/character/1',
  },
  {
    id: 2,
    name: 'Morty Smith',
    species: 'Human',
    gender: 'Male',
    created: '2017-11-04T18:50:21.651Z',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
      'https://rickandmortyapi.com/api/episode/3',
    ],
    image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    origin: {
      name: 'Earth (C-137)',
      url: 'https://rickandmortyapi.com/api/location/1',
    },
    status: 'Alive',
    type: '',
    url: 'https://rickandmortyapi.com/api/character/2',
  },
];

export const episodes = [
  'Rixty Minutes',
  'Something Ricked This Way Comes',
  'Close Rick-counters of the Rick Kind',
  'Ricksy Business',
  'A Rickle in Time',
];

export const locations = [
  {
    name: 'Citadel of Ricks',
    type: 'Space station',
    dimension: 'unknown',
  },
  {
    name: 'Earth (Replacement Dimension)',
    type: 'Planet',
    dimension: 'Replacement Dimension',
  },
];
