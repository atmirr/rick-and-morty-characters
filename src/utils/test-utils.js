import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

const mockedInitialState = {
  characters: { items: [], nextPage: null, error: null, isLoading: false },
  episodes: { items: [], nextPage: null, error: null, isLoading: false },
  locations: { items: [], error: null, isLoading: false },
};

function render(
  ui,
  {
    initialState = mockedInitialState,
    store = configureStore()(initialState),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
