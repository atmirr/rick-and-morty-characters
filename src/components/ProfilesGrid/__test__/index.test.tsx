import React from 'react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import { screen, render } from 'src/utils/test-utils';
import { characters } from 'src/utils/mocked-data';

import ProfilesGrid from '../index';

describe('<ProfilesGrid />', () => {
  it('should renders the component', () => {
    const { container } = render(<ProfilesGrid />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display profiles if the profiles prop has been passed', () => {
    const { container } = render(<ProfilesGrid profiles={characters} />);
    expect(screen.queryAllByTestId('profiles-grid').length).toBe(1);
    const imageNodes = container.querySelectorAll('img');
    expect(imageNodes.length).toBe(2);
    expect(screen.queryAllByText(/Rick Sanchez/i).length).toBe(1);
    expect(screen.queryAllByText(/Human/i).length).toBe(2);
    expect(screen.queryAllByText(/(Male)/i).length).toBe(2);
    expect(screen.queryAllByText(/Alive/i).length).toBe(2);
  });

  it('should render 20 profile placeholders when the profiles prop is empty and isLoading prop has been set to true', () => {
    render(<ProfilesGrid profiles={[]} isLoading />);
    expect(screen.queryAllByTestId('profiles-grid').length).toBe(1);
    expect(
      screen.queryAllByTestId(/^(profile-placeholder-)+([0-9]{1,2})$/).length,
    ).toBe(20);
  });

  it('should render 20 profile placeholders when the isLoading prop has been set to true and the profiles prop has been passed', () => {
    render(<ProfilesGrid profiles={characters} isLoading />);
    expect(screen.queryAllByTestId('profiles-grid').length).toBe(1);
    expect(
      screen.queryAllByTestId(/^(profile-placeholder-)+([0-9]{1,2})$/).length,
    ).toBe(20);
  });

  it('should display an error text when error prop has been passed', () => {
    const error = '404: Not found!';
    render(<ProfilesGrid error={error} />);
    expect(
      screen.queryAllByText(`There is something wrong! Error: ${error}`).length,
    ).toBe(1);
  });
});
