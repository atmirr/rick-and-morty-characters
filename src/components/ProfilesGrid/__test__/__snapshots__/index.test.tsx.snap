// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfilesGrid /> should renders the component 1`] = `
.c0 {
  max-width: 1400px;
  margin: 0 auto;
  padding-bottom: 40px;
}

.c1 {
  padding: 30px 25px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div
  class="c0"
  data-testid="profiles-grid"
>
  <div
    class="c1"
  />
</div>
`;
